phoenix-core/src/main/java/org/apache/phoenix/schema/MetaDataClient.java

```diff
@@ -945,55 +963,79 @@ public MutationState createIndex(CreateIndexStatement statement, byte[][] splits
-                    if (col.getViewConstant() == null) {
-                        PDataType dataType = IndexUtil.getIndexColumnDataType(col);
-                        colName = ColumnName.caseSensitiveColumnName(IndexUtil.getIndexColumnName(col));
-                        allPkColumns.add(new Pair<ColumnName, SortOrder>(colName, pair.getSecond()));
-                        columnDefs.add(FACTORY.columnDef(colName, dataType.getSqlTypeName(), col.isNullable(), col.getMaxLength(), col.getScale(), false, SortOrder.getDefault()));
-                    }
+                Set<ColumnName> indexedColumnNames = Sets.newHashSetWithExpectedSize(indexParseNodeAndSortOrderList.size());
+                for (Pair<ParseNode, SortOrder> pair : indexParseNodeAndSortOrderList) {
+                	ParseNode parseNode = pair.getFirst();
+            	    // compile the parseNode to get an expression
+                    PhoenixStatement phoenixStatment = new PhoenixStatement(connection);
+                    final StatementContext context = new StatementContext(phoenixStatment, resolver);
+                    // normalize the parse node
+                    parseNode = StatementNormalizer.normalize(parseNode, resolver);
+                    ExpressionCompiler expressionCompiler = new ExpressionCompiler(context);
+                    Expression expression = parseNode.accept(expressionCompiler);   
+                    if (expressionCompiler.isAggregate()) {
+                        throw new SQLExceptionInfo.Builder(SQLExceptionCode.AGGREGATE_EXPRESSION_NOT_ALLOWED_IN_INDEX).build().buildException();
+                    }
+                    if (expression.getDeterminism() != Determinism.ALWAYS) {
+                        throw new SQLExceptionInfo.Builder(SQLExceptionCode.NON_DETERMINISTIC_EXPRESSION_NOT_ALLOWED_IN_INDEX).build().buildException();
+                    }
+                    // true for any constant (including a view constant), as we don't need these in the index
+                    if (expression.isStateless()) {
+                        continue;
+                    }
+                    unusedPkColumns.remove(expression);
+                    
+                    ColumnName colName = null;
+					// if expression is KeyValueExpression set the column name correctly
+					if (expression.getChildren().size() == 0) {
+						colName = expression
+								.accept(new StatelessTraverseAllExpressionVisitor<ColumnName>() {
+									@Override
+									public ColumnName visit(
+											KeyValueColumnExpression node) {
+										return ColumnName.caseSensitiveColumnName(Bytes.toString(IndexUtil
+												.getIndexColumnName(node.getColumnFamily(), node.getColumnName())));
+									}
+								});
+					}
+					colName = colName!=null ? colName : ColumnName.caseSensitiveColumnName(IndexUtil.getIndexColumnName(null, expression.toString().replaceAll("\"", "")));
```

Comment:
I don't get this still. Why do you have to remove all double quotes? That could be problematic, for example if a string constant contains a double quote.